src/datasources/notifications/notifications.datasource.spec.ts

```diff
@@ -0,0 +1,633 @@
+                return {
+                  id: expect.any(Number),
+                  notification_subscription_id: i + 1,
+                  notification_type_id: types.find((t) => t.name === type)?.id,
+                };
+              });
+            }),
+          ),
+        );
+      });
+    });
+
+    it('should always update the deviceType/cloudMessagingToken', async () => {
+      const signerAddress = getAddress(faker.finance.ethereumAddress());
+      const upsertSubscriptionsDto = upsertSubscriptionsDtoBuilder().build();
+      const secondSubscriptionsDto = upsertSubscriptionsDtoBuilder()
+        .with('deviceUuid', upsertSubscriptionsDto.deviceUuid)
+        .build();
+      await target.upsertSubscriptions({
+        signerAddress,
+        upsertSubscriptionsDto,
+      });
+
+      // Insert should not throw despite it being the same device UUID
+      await expect(
+        target.upsertSubscriptions({
+          signerAddress,
+          upsertSubscriptionsDto: secondSubscriptionsDto,
+        }),
+      ).resolves.not.toThrow();
+      // Device UUID should have updated
+      await expect(
+        sql`SELECT * FROM push_notification_devices`,
+      ).resolves.toStrictEqual([
+        {
+          id: 1,
+          device_type: secondSubscriptionsDto.deviceType,
+          device_uuid: expect.any(String),
+          cloud_messaging_token: secondSubscriptionsDto.cloudMessagingToken,
+          created_at: expect.any(Date),
+          updated_at: expect.any(Date),
+        },
+      ]);
+    });
+
+    it('should update a subscription, setting only the newly subscribed notification types', async () => {
+      const signerAddress = getAddress(faker.finance.ethereumAddress());
+      const upsertSubscriptionsDto = upsertSubscriptionsDtoBuilder()
+        .with('safes', [
+          {
+            chainId: faker.string.numeric(),
+            address: getAddress(faker.finance.ethereumAddress()),
+            notificationTypes: faker.helpers.arrayElements(
+              Object.values(NotificationType),
+            ),
+          },
+        ])
+        .build();
+      const newNotificationTypes = faker.helpers.arrayElements(
+        Object.values(NotificationType),
+      );
+      await target.upsertSubscriptions({
+        signerAddress,
+        upsertSubscriptionsDto,
+      });
+      await target.upsertSubscriptions({
+        signerAddress,
+        upsertSubscriptionsDto: {
+          ...upsertSubscriptionsDto,
+          safes: [
+            {
+              ...upsertSubscriptionsDto.safes[0],
+              notificationTypes: newNotificationTypes,
+            },
+          ],
+        },
+      });
+
+      await Promise.all([
+        sql`SELECT * FROM notification_types`,
+        sql`SELECT * FROM notification_subscription_notification_types`,
+      ]).then(([notificationTypes, subscribedNotifications]) => {
+        // Only new notification types should be subscribed to
```

Comment:
IFAIU the notification types are replaced by the new ones completely (no nested upsert). I think this is correct, but I'd make it more explicit in the test by setting 2 or more items in `upsertSubscriptionsDto.safes`, and then check `subscribedNotifications` was overwritten (length is 1 and the only item is the expected one).

Reply:
Adjusted in d49aca2e6a097d8ac438fa5876d39f1b2006bf3f.